core/dropdowndiv.js

```diff
@@ -0,0 +1,519 @@
+  // for the animated transition to final X, Y is a workaround.
+
+  Blockly.DropDownDiv.positionInternal_(
+      metrics.initialX, metrics.initialY,
+      metrics.finalX, metrics.finalY);
+  return metrics.arrowAtTop;
+};
+
+/**
+ * Helper to position the drop-down and the arrow, maintaining bounds.
+ * See explanation of origin points in Blockly.DropDownDiv.show.
+ * @param {number} primaryX Desired origin point x, in absolute px
+ * @param {number} primaryY Desired origin point y, in absolute px
+ * @param {number} secondaryX Secondary/alternative origin point x, in absolute px
+ * @param {number} secondaryY Secondary/alternative origin point y, in absolute px
+ * @returns {Object} Various final metrics, including rendered positions for drop-down and arrow.
+ */
+Blockly.DropDownDiv.getPositionMetrics = function(primaryX, primaryY, secondaryX, secondaryY) {
+  var div = Blockly.DropDownDiv.DIV_;
+  var boundPosition = Blockly.DropDownDiv.boundsElement_.getBoundingClientRect();
+
+  var boundSize = goog.style.getSize(Blockly.DropDownDiv.boundsElement_);
+  var divSize = goog.style.getSize(div);
+
+  // First decide if we will render at primary or secondary position
+  // i.e., above or below
+  // renderX, renderY will eventually be the final rendered position of the box.
+  var renderX, renderY, renderedSecondary, renderedTertiary;
+  // Can the div fit inside the bounds if we render below the primary point?
+  if (primaryY + divSize.height > boundPosition.top + boundSize.height) {
+    // We can't fit below in terms of y. Can we fit above?
+    if (secondaryY - divSize.height < boundPosition.top) {
+      // We also can't fit above, so just render at the top of the screen.
+      renderX = primaryX;
+      renderY = 0;
+      renderedSecondary = false;
+      renderedTertiary = true;
+    } else {
+      // We can fit above, render secondary
+      renderX = secondaryX;
+      renderY = secondaryY - divSize.height - Blockly.DropDownDiv.PADDING_Y;
+      renderedSecondary = true;
+    }
+  } else {
+    // We can fit below, render primary
+    renderX = primaryX;
+    renderY = primaryY + Blockly.DropDownDiv.PADDING_Y;
+    renderedSecondary = false;
+  }
+  // First calculate the absolute arrow X
+  // This needs to be done before positioning the div, since the arrow
+  // wants to be as close to the origin point as possible.
+  var arrowX = renderX - Blockly.DropDownDiv.ARROW_SIZE / 2;
+  // Keep in overall bounds
+  arrowX = Math.max(boundPosition.left, Math.min(arrowX, boundPosition.left + boundSize.width));
+
+  // Adjust the x-position of the drop-down so that the div is centered and within bounds.
+  var centerX = divSize.width / 2;
+  renderX -= centerX;
+  // Fit horizontally in the bounds.
+  renderX = Math.max(
+      boundPosition.left,
+      Math.min(renderX, boundPosition.left + boundSize.width - divSize.width)
+  );
+  // After we've finished caclulating renderX, adjust the arrow to be relative to it.
+  arrowX -= renderX;
```

Comment:
This is just removing renderX from the original positioning of the arrow, but after things have been adjusted. You might as well just set arrowX = -ARROW_SIZE/2. I think the ordering in this method is all wrong.